File: app/src/app/models/tarot-zodiac.ts
import { TarotAstrology } from './tarot-astrology';
import { TarotElement } from './tarot-element';
import { TarotModality } from './tarot-modality';
import { TarotPlanet } from './tarot-planet';

export interface TarotZodiac extends TarotAstrology {
    elementId: number;
    modalityId: number;
    rulerId: number;
    dates: string;
    positiveTraits: string[];
    negativeTraits: string[];

    element: TarotElement;
    modality: TarotModality;
    ruler: TarotPlanet;
}

File: server/Tarot.Models/Models/TarotZodiac.cs
namespace Tarot.Models;

public class TarotZodiac : TarotAstrology
{
    private readonly DateOnly StartDate;
    private readonly DateOnly EndDate;
    public int ElementId { get; }
    public int ModalityId { get; }
    public int RulerId { get; }
    public override string Name { get; }
    public string Dates => $"{StartDate:MMM dd} - {EndDate:MMM dd}";
    public string[] PositiveTraits { get; set; }
    public string[] NegativeTraits { get; set; }

    public TarotZodiac(
        int id,
        string name,
        DateOnly startDate,
        DateOnly endDate,
        TarotElement element,
        TarotModality modality,
        TarotPlanet ruler,
        string[] positiveTraits,
        string[] negativeTraits
    ) : base(id, "Zodiac")
    {
        Name = name;
        StartDate = startDate;
        EndDate = endDate;
        ElementId = element.Id;
        ModalityId = modality.Id;
        RulerId = ruler.Id;
        PositiveTraits = positiveTraits;
        NegativeTraits = negativeTraits;
    }

    public TarotElement Element => TarotElements.Elements.Get(ElementId);
    public TarotModality Modality => TarotModalities.Modalities.Get(ModalityId);
    public TarotPlanet Ruler => TarotPlanets.Planets.Get(RulerId);

    public override string[] Keywords =>
        PositiveTraits
            .Concat(NegativeTraits)
            .ToArray();
}
